doc/syscalls/10001_ros.md

```diff
@@ -0,0 +1,88 @@
+---
+driver number: 0x10001
+---
+
+# Read Only Syscalls
+
+## Overview
+
+This capsule provides read only system calls to userspace applications.
+This is similar to the Linux vDSO syscalls.
+
+The benefit of using these is that applications can avoid the context
+switch overhead of traditional syscalls by just reading the value from
+memory.
+
+The value will only be as accurate as the last time the application was
+switched to by the kernel.
+
+The layout of the read only syscalls in the allow region depends on the
+version. Userspace can use `command 0` to get the version information.
+
+Versions are backwards compatible, that is new versions will only add
+fields, not remove existing ones or change the order.
+
+Version 1:
+  |-------------------------|
+  |    Switch Count (u32)   |
+  |-------------------------|
+  |   Pending Tasks (u32)   |
+  |-------------------------|
+  |                         |
+  |     Time Ticks (u64)    |
+  |-------------------------|
+
+Example C code to safely read the 64-bit timer value is included below,
+where `ptr` is a `uint32_t*` pointing to the ROS memory region.
+
+```c
+  volatile uint32_t high, low;
+
+  do {
+    // Set the high bytes the value in memory
+    time_ticks_high = ptr[3];
+    // Read the low bytes
+    time_ticks_low = ptr[2];
+    // If the high bytes don't match what is still in memory re-try
+    // the load
+  } while (time_ticks_high != ptr[3]);
+```
```

Comment:
```suggestion
```c
  volatile uint32_t* ptr; // Points to ROS memory region
  uint32_t time_ticks_high, time_ticks_low;
  uint32_t switch_count = 0;

  do {
    // Start our attempt at an atomic read
    switch_count = ptr[0];
    // Set the high bytes the value in memory
    time_ticks_high = ptr[3];
    // Read the low bytes
    time_ticks_low = ptr[2];
    // Check if there was a context switch while reading this value out, if so, retry
  } while (switch_count != ptr[0]);
```
```

I think this is a more accurate code snippet to read according to the mechanisms outlined here?